x-pack/plugins/apm/typings/common.ts

```diff
@@ -7,3 +7,9 @@
 export interface StringMap<T = unknown> {
   [key: string]: T;
 }
+
+export type PromiseReturnType<T> = T extends Promise<infer U>
+  ? U
+  : T extends (...args: any[]) => Promise<infer V>
+  ? V
+  : T;
```

Comment:
This is awesome!

Reply:
Glad you like it :)

Reply:
... I think it's pretty Awesome too. These types of generics used to scare the hell out of me, but I like them in small doses :D